File: training task/polymorphism_runtime.py
class Student:
    def __init__(self, name):
        self.name = name

    def display_details(self):
        print(f"Name: {self.name}")

class UndergraduateStudent(Student):
    def __init__(self, name, year):
        super().__init__(name)
        self.year = year

    def display_details(self):  #derived function from base
        print("Undergraduate Student Details:")
        super().display_details()  
        print(f"Year: {self.year}")

class GraduateStudent(Student):
    def __init__(self, name, thesis_title):
        super().__init__(name)
        self.thesis_title = thesis_title

    def display_details(self):   #derived function from base
        print("Graduate Student Details:")
        super().display_details()
        print(f"Thesis Title: {self.thesis_title}")


undergrad = UndergraduateStudent("Nitu", 3)
grad = GraduateStudent("Mou", "Machine Learning in Healthcare")

print("Displaying Undergraduate Student Details:")
undergrad.display_details()

print("\nDisplaying Graduate Student Details:")
grad.display_details()

File: training task/polymorphism_runtime.cpp
// #include <iostream>
// #include <string>
// using namespace std;


// class Student {
// protected:
//     string name;

// public:
//     Student(string name) {

//         this->name=name;
//     }
//     virtual void displayDetails() { 
//         cout << "Name: " << name << endl; 
//     }
// };

// class UndergraduateStudent : public Student {
// private:
//     int year;

// public:
//     UndergraduateStudent(const string& name, int year) : Student(name){
//         this->year=year;
//     }

   
//     void displayDetails() override {
//         cout << "Undergraduate Student Details:" << endl;
//         Student::displayDetails(); 
//         cout << "Year: " << year << endl; 
//     }
// };


// class GraduateStudent : public Student {
// private:
//     string thesisTitle;

// public:
//     GraduateStudent(const string& name, const string& thesisTitle) : Student(name) {
//         this->thesisTitle=thesisTitle;
//     }

   
//     void displayDetails() override {
//         cout << "Graduate Student Details:" << endl;
//         Student::displayDetails(); 
//         cout << "Thesis Title: " << thesisTitle << endl;
//     }
// };

// int main() {

//     UndergraduateStudent undergrad("Mohua", 3);
//     GraduateStudent grad("Sadia", "Machine Learning in Healthcare");
//     cout << "Displaying Undergraduate Student Details:" << endl;
//     undergrad.displayDetails();
//     cout << endl;
//     cout << "Displaying Graduate Student Details:" << endl;
//     grad.displayDetails();

//     return 0;
// }


